core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorConcurrencyTest.scala

```diff
@@ -106,25 +97,56 @@ class GroupCoordinatorConcurrencyTest extends AbstractCoordinatorConcurrencyTest
     }
   }
 
-  def createGroupMembers(groupPrefix: String): Set[GroupMember] = {
-    (0 until nGroups).flatMap { i =>
-      new Group(s"$groupPrefix$i", nMembersPerGroup, groupCoordinator, replicaManager).members
-    }.toSet
+  private var producerIdCount = 0L
+  private val allGroupMembers = mutable.ArrayBuffer[GroupMember]()
+
+  def groupMembers(groupId: String, nMembers: Int, groupCoordinator: GroupCoordinator): Seq[GroupMember] = {
+    val groupPartitionId = groupCoordinator.partitionFor(groupId)
+    groupCoordinator.groupManager.addPartitionOwnership(groupPartitionId)
+    val lock = new ReentrantLock()
+    val producerId = producerIdCount
+    producerIdCount += 1
+    val members = (0 until nMembers).map(i => new GroupMember(groupId = groupId,
+      groupPartitionId = groupPartitionId,
+      leader = i == 0,
+      producerId = producerId,
+      txnLock = lock))
+    allGroupMembers ++= members
+    members
   }
 
+  def createGroupMembers(groupPrefix: String): Set[GroupMember] =
+    (0 until nGroups).flatMap(i => groupMembers(s"$groupPrefix$i", nMembersPerGroup, groupCoordinator)).toSet
+
   @Test
   def testConcurrentGoodPathSequence(): Unit = {
     verifyConcurrentOperations(createGroupMembers, allOperations)
   }
 
   @Test
   def testConcurrentTxnGoodPathSequence(): Unit = {
-    verifyConcurrentOperations(createGroupMembers, allOperationsWithTxn)
+    verifyConcurrentOperations(createGroupMembers, Seq(
+      new JoinGroupOperation,
+      new SyncGroupOperation,
+      new OffsetFetchOperation,
+      new CommitTxnOffsetsOperation(needLock = false),
+      new CompleteTxnOperation(needLock = false),
```

Comment:
Hmm, why don't we need the lock here since CommitTxnOffsetsOperation and CompleteTxnOperation could still run in parallel?

Reply:
you are right